<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta http-equiv="X-UA-Compatible" content="ie=edge">
  <title>07_Promise的API</title>
</head>
<body>
  <!-- 
    1. Promise构造函数: Promise (excutor) {}
        excutor函数: 同步执行  (resolve, reject) => {}
        resolve函数: 内部定义成功时我们调用的函数 value => {}
        reject函数: 内部定义失败时我们调用的函数 reason => {}
        说明: excutor会在Promise内部立即同步回调,异步操作在执行器中执行

    2. Promise.prototype.then方法: (onResolved, onRejected) => {}
        onResolved函数: 成功的回调函数  (value) => {}
        onRejected函数: 失败的回调函数 (reason) => {}
        说明: 指定用于得到成功value的成功回调和用于得到失败reason的失败回调
              返回一个新的Promise实例对象

    3. Promise.prototype.catch方法: (onRejected) => {}
        onRejected函数: 失败的回调函数 (reason) => {}
        说明: then()的语法糖, 相当于: then(undefined, onRejected)

    4. Promise.resolve方法: (value) => {}
        value: 成功的数据或promise对象
        说明: 返回一个成功的promise对象

    5. Promise.reject方法: (reason) => {}
        reason: 失败的原因
        说明: 返回一个失败的promise对象

    6. Promise.all方法: (promises) => {}
        promises: 包含n个promise的数组
        说明: 返回一个新的promise, 只有所有的promise都成功才成功, 只要有一个失败了就直接失败

    7. Promise.race方法: (promises) => {}
        promises: 包含n个promise的数组
        说明: 返回一个新的promise, 第一个完成的promise的结果状态就是最终的结果状态
  -->

<script type="text/javascript">

  //catch的使用
  /*new Promise((resolve,reject)=>{
    setTimeout(()=>{
      reject(1)
    },1000)
  }).catch(
    reason => console.log('onRejected',reason)
  )*/

  //Promise.resolve()和Promise.reject()
  /*let p1 = new Promise((resolve,reject)=>{
    reject(1)
  })
  let p2 = Promise.reject(2)
  let p3 = Promise.resolve(3)
  console.log(p3)*/

  //Promise.all和Promise.race
  /*let p1 = new Promise((resolve,reject)=>{
    setTimeout(()=>{
      reject(1)
    },1000)
  })
  let p2 = Promise.resolve(2)
  let p3 = Promise.resolve(3)

  //Promise.all
  let p4 = Promise.all([p1,p2,p3])
  p4.then(
    values => console.log('成功的回调用了',values),
    reason => console.log('失败的回调用了',reason)
  )

  //Promise.race
  let p5 = Promise.race([p3,p1,p2])
  p5.then(
    value => console.log('成功的回调用了',value),
    reason => console.log('失败的回调用了',reason)
  )*/


</script>
</body>
</html>